File: mill-example/oystercard/src/oystercard/java/TubeStation.java
package oystercard.java;

import java.util.HashSet;
import java.util.Set;

/**
 * Contains our static reference data for all known oystercard stations and a means to find a station based on its name
 */
public enum TubeStation {

    EarlsCourt("Earl’s Court", 1),
    Holborn("Holborn", 1, 2),
    Wimbledon("Wimbledon", 3),
    Hammersmith("Hammersmith", 2);

    public final Set<Zone> zones;
    public final String stationName;

    TubeStation(String name, int firstZone, int ... theRest) {
        Set<Zone> zoneSet = new HashSet<>();
        zoneSet.add(new Zone(firstZone  ));
        for (int z : theRest) {
            zoneSet.add(new Zone(z));
        }
        stationName = name;

        zones = java.util.Collections.unmodifiableSet(zoneSet);
    }
}

File: mill-example/oystercard/src/oystercard/scala/TubeStation.scala
package oystercard.scala

/**
  * @param name  the station name
  * @param zones the zone(s) the station is in
  */
class TubeStation private[oystercard](val name: String, zoneIds: Int*) {
  require(zoneIds.nonEmpty)
  val zones: Set[Zone] = zoneIds.toSet
}

/**
  * Contains our static reference data for all known oystercard stations and a means to find a station based on its name
  */
object TubeStation {
  def apply(station: String, firstZone: Int, theRest: Int*): TubeStation = {
    new TubeStation(station, (firstZone +: theRest): _*)
  }

  def unapply(station: String): Option[TubeStation] = stations.find(_.name == station)

  case object EarlsCourt extends TubeStation("Earl’s Court", 1)

  case object Holborn extends TubeStation("Holborn", 1, 2)

  case object Wimbledon extends TubeStation("Wimbledon", 3)

  case object Hammersmith extends TubeStation("Hammersmith", 2)

  def stations = Set(EarlsCourt, Holborn, Wimbledon, Hammersmith)
}
